### ISSUE TITLE
Question: How do you handle Runner errors?

### ISSUE THREAD

**Issue by orlu-nd** [dunglas/frankenphp#961]
First of all, we absolutely love FrankenPHP. It handles our Symfony applications beautifully.

However, we're at our wit's end trying to catch Runner errors, like the one shown in this example:

```bash
PHP Warning:  POST Content-Length of 10152058 bytes exceeds the limit of 5242880 bytes in /app/vendor/runtime/frankenphp-symfony/src/Runner.php on line 45	{"syslog_level": "warning"}
```

In `prod` mode, the example above is logged as a 'Warning'. However, in `dev` mode, it results in a PHP Fatal Error:

```bash
ERROR	PHP Fatal error:  Uncaught ErrorException: Warning: POST Content-Length of 10152058 bytes exceeds the limit of 5242880 bytes in /app/vendor/runtime/frankenphp-symfony/src/Runner.php:45
```

When using FrankenPHP and Caddy together, the request is indeed blocked because the Content-Length exceeds the PHP configuration limits set in `php.ini`:

```ini
; max post limit:
upload_max_filesize = 4M
post_max_size = 5M
```

From a user experience standpoint, however, the visitor receives a "The connection was reset" error (NS_ERROR_NET_RESET). This is what we are trying to solve. We would like to present a nice error page.

We have explored the following options, but without success:

-  Attempting to catch any type of error in Runner.php, index.php, or Kernel.php.
-  Using [handle_errors](https://caddyserver.com/docs/caddyfile/directives/handle_errors) in our Caddy configuration.

How are we supposed to handle this situation in a proper manner? Once again, our application runs great, and everything else, including errors thrown by Symfony, is handled perfectly. We're just struggling with this minor edge case and would appreciate any guidance on how to properly manage it. Thank you in advance for any assistance.

Caddyfile and Dockerfile listed below.

Our `Caddyfile`:

```
{
	{$CADDY_GLOBAL_OPTIONS}

	frankenphp {
	    worker {
	        file ./public/index.php
	        num 10
	        env APP_RUNTIME "Runtime\FrankenPhpSymfony\Runtime"
	    }

	}

	order php_server before file_server
}

{$CADDY_EXTRA_CONFIG}

{$SERVER_NAME:localhost} {

    header {
        # Only allow GET and POST.
        +Access-Control-Allow-Methods "GET, POST"

        # Set Content-Security-Policy
        +Content-Security-Policy "default-src 'self';"

        # Set Cross-Origin-Embedder-Policy
        +Cross-Origin-Embedder-Policy "require-corp"

        # Set Cross-Origin-Opener-Policy
        +Cross-Origin-Opener-Policy "same-origin"

        # Set Cross-Origin-Resource-Policy
        +Cross-Origin-Resource-Policy "same-origin"

        # Set Strict-Transport-Security
        +Strict-Transport-Security "max-age=31536000"

        # Set X-XSS-Protection
        +X-XSS-Protection "1; mode=block"

        # Set X-Content-Type-Options
        +X-Content-Type-Options "nosniff"

        # Set X-XSS-Protection
        +X-Frame-Options "SAMEORIGIN"

        # Set Referrer-Policy
        +Referrer-Policy "strict-origin"

        # Set Permissions-Policy
        +Permissions-Policy "fullscreen=()"

        # Delete the X-Generator header that is set by the Sulu CMS:
        # https://github.com/sulu/sulu-standard/blob/master/CHANGELOG.md
        # #2192 [WebsiteBundle] Added X-Generator HTTP header for Sulu website detection
        -X-Generator

        # Delete the Server header
        -Server
    }

   root * public/
   encode zstd br gzip

   {$CADDY_SERVER_EXTRA_DIRECTIVES}

   php_server
}
```

Our `Dockerfile`:

```
FROM dunglas/frankenphp:latest-php8.3-alpine

# add additional extensions here:
RUN install-php-extensions \
	pdo_pgsql \
	intl \
	redis \
	sysvsem \
	opcache

RUN mv "$PHP_INI_DIR/php.ini-production" "$PHP_INI_DIR/php.ini"

COPY --link .docker/frankenphp/conf.d/app.ini $PHP_INI_DIR/conf.d/
COPY --link .docker/frankenphp/Caddyfile /etc/caddy/Caddyfile

COPY . /app
```

**Comment by sirpilan** [comment]
We have a similar problem, where Exceptions get catched fine, but **Throwables** are not catched. Only in Worker-Mode.

**Comment by withinboredom** [comment]
> In prod mode, the example above is logged as a 'Warning'. However, in dev mode, it results in a PHP Fatal Error

A warning isn't an exception, so you can't "catch" them. In this case, it is probably better to check the file array for errors or turn the warning into an exception in production.

**Comment by sirpilan** [comment]
> We have a similar problem, where Exceptions get catched fine, but **Throwables** are not catched. Only in Worker-Mode.

Found whats wrong. We upgraded from Symfony 5 to 6, where a new framework parameter was introduced `handle_all_throwables`. In 5 throwables get catched by default in 6 it has to be enabled explicitly (even though it gets catched in non-worker mode). In 7 this will be defaulted to true. In worker mode on symfony 6 however the Throwables dont get catched (if you omit `handle_throwables`), so they dont get logged (but are visible in docker logs). Maybe helps if you have a look at your config - did the trick for us.

Would still love to know, why there is a different behaviour in worker/non-worker mode.

**Comment by orlu-nd** [comment]
@sirpilan Thank you for your suggestions. I have checked the `handle_all_throwables`, it is set to true in our application. We are on Symfony 7. 

```bash
php bin/console debug:config framework | grep handle_all_throwables
handle_all_throwables: true
```

Changing this setting (just trying) did not make any difference. 

As for your other suggestion; worker/non-worker mode. We have a similar issue in non-worker mode.

@withinboredom Thank you for your suggestion.

You are indeed correct; 'Warnings' are not the same as exceptions, however in production mode, the following happens:

```
Warning: POST Content-Length of 10152047 bytes exceeds the limit of 5242880 bytes in /app/vendor/runtime/frankenphp-symfony/src/Runner.php on line 45

Fatal error: Uncaught RuntimeException: Failed to start the session because headers have already been sent by "/app/vendor/runtime/frankenphp-symfony/src/Runner.php" at line 45
```

Looking at the $_FILES or $_SERVER['CONTENT-LENGHT'] is something we did not think of. Thank you. However, due to the all the Error/Warnings, the headers are already sent. Therefore, we are unable to generate our own error message. Setting "display_errors" to 'Off' (or 0) results in FrankenPHP sending a 0byte to the browser and NS_ERROR_NET_RESET.

To both: I will create an example application where I can demonstrate the issue.

**Comment by orlu-nd** [comment]
Here is an example application to demonstrate the problem:
https://github.com/orlu-nd/frankenphp-runner-example

Once again, thank you in advance for any assistance.

**Comment by dunglas** [comment]
Unfortunately, it looks like there is no way to disable this specific warning. [According to the docs](https://www.php.net/manual/en/ini.core.php#ini.post-max-size), when this case occurs, `$_POST` and `$_GET` are empty.

I think that two alternative solutions can be used as workarounds:

1. Handle requests that are too big at the Caddy level. You can try something like this (I haven't tried):
    ```caddyfile
    @toobig expression {http.request.header.Content-Length} != "" && int({http.request.header.Content-Length}) > 9999
    redir @toobig /my-error-page 302
    # FrankenPHP directives
    ```
2. Disable warnings reporting using `error_reporting` in  `php.ini`, check if `$_POST` is empty and generate an error in this case

In my opinion, solution 1 will be more reliable and is cleaner (PHP isn't called at all if the request is too big, which is better).

**Comment by orlu-nd** [comment]
Thank you, @dunglas, for your suggestion. It's an approach we hadn't considered before. Unfortunately, using `redir` continued to result in the same `NS_ERROR_NET_RESET` issue. However, by building on your approach, we were able to find a solution with the following configuration:

```Caddyfile
@toobig expression {http.request.header.Content-Length} != "" && int({http.request.header.Content-Length}) > 9999
error @toobig "Entity too large" 413

handle_errors {
    rewrite * /error.html
    file_server
}
```

The inclusion of `file_server` proved to be crucial. Omitting it or using `php_server` instead kept triggering the `NS_ERROR_NET_RESET` issue, likely because the redirect was still being processed by PHP/FrankenPHP. While it may not be the cleanest solution, we're satisfied with the outcome for this particular edge case.

Thank you all for your help.

**Comment by dunglas** [comment]
Thanks for providing your working solution @orlu-nd. By the way `9999` must be replaced by the value of your `max_post_size` config in `php.ini`.

That would be nice to include this in the documentation. Would you mind doing it? If you don't have time, can I reuse your snippet? Thanks!

**Comment by orlu-nd** [comment]
Certainly, I'd be happy to contribute. Would it be appropriate to add it under 'Known issues'?
